In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('bookings.csv', sep=';')

In [3]:
df.head(10)

Hotel  Is Canceled  Lead Time arrival full date  Arrival Date Year  \
0  Resort Hotel            0        342        2015-07-01               2015   
1  Resort Hotel            0        737        2015-07-01               2015   
2  Resort Hotel            0          7        2015-07-01               2015   
3  Resort Hotel            0         13        2015-07-01               2015   
4  Resort Hotel            0         14        2015-07-01               2015   
5  Resort Hotel            0         14        2015-07-01               2015   
6  Resort Hotel            0          0        2015-07-01               2015   
7  Resort Hotel            0          9        2015-07-01               2015   
8  Resort Hotel            1         85        2015-07-01               2015   
9  Resort Hotel            1         75        2015-07-01               2015   

  Arrival Date Month  Arrival Date Week Number  Arrival Date Day of Month  \
0               July                        27                          1   
1               July                        27                          1   
2               July                        27                          1   
3               July                        27                          1   
4               July                        27                          1   
5               July                        27                          1   
6               July                        27                          1   
7               July                        27                          1   
8               July                        27                          1   
9               July                        27                          1   

   Stays in Weekend nights  Stays in week nights  ...  Adults  Children  \
0                        0                     0  ...       2       0.0   
1                        0                     0  ...       2       0.0   
2                        0                     1  ...       1       0.0   
3                        0                     1  ...       1       0.0   
4                        0                     2  ...       2       0.0   
5                        0                     2  ...       2       0.0   
6                        0                     2  ...       2       0.0   
7                        0                     2  ...       2       0.0   
8                        0                     3  ...       2       0.0   
9                        0                     3  ...       2       0.0   

   Babies  Meal Country Reserved Room Type Assigned room type customer type  \
0       0    BB     PRT                  C                  C     Transient   
1       0    BB     PRT                  C                  C     Transient   
2       0    BB     GBR                  A                  C     Transient   
3       0    BB     GBR                  A                  A     Transient   
4       0    BB     GBR                  A                  A     Transient   
5       0    BB     GBR                  A                  A     Transient   
6       0    BB     PRT                  C                  C     Transient   
7       0    FB     PRT                  C                  C     Transient   
8       0    BB     PRT                  A                  A     Transient   
9       0    HB     PRT                  D                  D     Transient   

  Reservation Status Reservation status_date  
0          Check-Out              2015-07-01  
1          Check-Out              2015-07-01  
2          Check-Out              2015-07-02  
3          Check-Out              2015-07-02  
4          Check-Out              2015-07-03  
5          Check-Out              2015-07-03  
6          Check-Out              2015-07-03  
7          Check-Out              2015-07-03  
8           Canceled              2015-05-06  
9           Canceled              2015-04-22  

[10 rows x 21 columns]

Analyze hotel booking data. Description:
*    `Hotel` - type of hotel (City Hotel or Resort Hotel)
*    `Is canceled` - booking has been canceled (1) or not (0); not canceled is considered successful
*    `Lead time` - number of days between booking date and arrival date
*    `Arrival full date` - full arrival date
*    `Arrival date year` - year of arrival
*    `Arrival date month` - month of arrival
*    `Arrival date week number` - week of arrival
*    `Arrival date day of month` - day of arrival
*    `Stays in weekend nights` - the number of weekends (Saturday or Sunday) that the guest has booked for hotel stay
*    `Stays in week nights` - the number of days (Monday to Friday) that the guest has booked for stay at the hotel
*    `Stays total nights` - total number of nights booked (sum of two previous columns)
*    `Adults` - number of adults
*    `Children` - number of children
*    `Babies` - the number of babies
*    `Meal` - selected type of food
*    `Country` - the country of origin of the client
*    `Reserved room type` - type of reserved room
*    `Assigned room type` - the type of room received (may differ from the booked)
*    `Customer type` - type of booking
*    `Reservation status` - the value of the last reservation status: Canceled - was canceled by the client; Check-Out - the client checked in but has already left the hotel; No-Show - the client did not register and told the hotel the reason
*    `Reservation status date` - date of status update

In [4]:
df.shape

(119390, 21)

In [5]:
df.dtypes

Hotel                         object
Is Canceled                    int64
Lead Time                      int64
arrival full date             object
Arrival Date Year              int64
Arrival Date Month            object
Arrival Date Week Number       int64
Arrival Date Day of Month      int64
Stays in Weekend nights        int64
Stays in week nights           int64
stays total nights             int64
Adults                         int64
Children                     float64
Babies                         int64
Meal                          object
Country                       object
Reserved Room Type            object
Assigned room type            object
customer type                 object
Reservation Status            object
Reservation status_date       object
dtype: object

Bring the column names to a generic view (lowercase and replace the spaces with a lower underscore):

In [6]:
df.rename(columns=lambda x: x.replace(' ', '_').lower(), inplace=True)

In [7]:
df.head(10)

hotel  is_canceled  lead_time arrival_full_date  arrival_date_year  \
0  Resort Hotel            0        342        2015-07-01               2015   
1  Resort Hotel            0        737        2015-07-01               2015   
2  Resort Hotel            0          7        2015-07-01               2015   
3  Resort Hotel            0         13        2015-07-01               2015   
4  Resort Hotel            0         14        2015-07-01               2015   
5  Resort Hotel            0         14        2015-07-01               2015   
6  Resort Hotel            0          0        2015-07-01               2015   
7  Resort Hotel            0          9        2015-07-01               2015   
8  Resort Hotel            1         85        2015-07-01               2015   
9  Resort Hotel            1         75        2015-07-01               2015   

  arrival_date_month  arrival_date_week_number  arrival_date_day_of_month  \
0               July                        27                          1   
1               July                        27                          1   
2               July                        27                          1   
3               July                        27                          1   
4               July                        27                          1   
5               July                        27                          1   
6               July                        27                          1   
7               July                        27                          1   
8               July                        27                          1   
9               July                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  ...  adults  children  \
0                        0                     0  ...       2       0.0   
1                        0                     0  ...       2       0.0   
2                        0                     1  ...       1       0.0   
3                        0                     1  ...       1       0.0   
4                        0                     2  ...       2       0.0   
5                        0                     2  ...       2       0.0   
6                        0                     2  ...       2       0.0   
7                        0                     2  ...       2       0.0   
8                        0                     3  ...       2       0.0   
9                        0                     3  ...       2       0.0   

   babies  meal country reserved_room_type assigned_room_type customer_type  \
0       0    BB     PRT                  C                  C     Transient   
1       0    BB     PRT                  C                  C     Transient   
2       0    BB     GBR                  A                  C     Transient   
3       0    BB     GBR                  A                  A     Transient   
4       0    BB     GBR                  A                  A     Transient   
5       0    BB     GBR                  A                  A     Transient   
6       0    BB     PRT                  C                  C     Transient   
7       0    FB     PRT                  C                  C     Transient   
8       0    BB     PRT                  A                  A     Transient   
9       0    HB     PRT                  D                  D     Transient   

  reservation_status reservation_status_date  
0          Check-Out              2015-07-01  
1          Check-Out              2015-07-01  
2          Check-Out              2015-07-02  
3          Check-Out              2015-07-02  
4          Check-Out              2015-07-03  
5          Check-Out              2015-07-03  
6          Check-Out              2015-07-03  
7          Check-Out              2015-07-03  
8           Canceled              2015-05-06  
9           Canceled              2015-04-22  

[10 rows x 21 columns]

***
Which countries have the guests made the largest number of successful reservations from? List the top 5.

In [8]:
df \
    .query('is_canceled == 0') \
    .value_counts(subset=['country']) \
    .head(5)

country
PRT        21071
GBR         9676
FRA         8481
ESP         6391
DEU         6069
dtype: int64

***
How many nights on average are hotels of different types reserved?

In [9]:
df \
    .groupby('hotel') \
    .aggregate({'stays_total_nights': 'mean'})

stays_total_nights
hotel                           
City Hotel              2.978142
Resort Hotel            4.318547

***
Sometimes the type of room received by the client (assigned_room_type) differs from the originally booked (reserved_room_type). This can happen, for example, because of overbooking. How many similar observations have met in the dataset?

In [10]:
df[df['reserved_room_type'] != df['assigned_room_type']].shape[0]

14917

***
Analyze the dates of scheduled arrival:
* Which month were the largest number of successful bookings made in 2016? Has the most popular month in 2017 changed?
* Group by year and check which month the City Hotel type bookings were canceled for most times during each period?

In [11]:
df \
    .query('arrival_date_year == 2016 and is_canceled == 0') \
    .value_counts(['arrival_date_month']) 

arrival_date_month
October               3689
May                   3563
September             3372
April                 3367
March                 3347
August                3238
June                  3196
July                  3073
November              2818
February              2554
December              2462
January               1691
dtype: int64

In [12]:
df \
    .query('arrival_date_year == 2017 and is_canceled == 0') \
    .value_counts(['arrival_date_month'])

arrival_date_month
May                   3551
July                  3329
March                 3298
June                  3208
April                 3198
August                3109
February              2818
January               2431
dtype: int64

`October` was the most productive in terms of successful reservations in `2016`. 
In `2017`, the situation has changed slightly, we see the leading positions in the `spring` months (although the figures for them are about the same as in 2016)

In [12]:
df \
    .query('hotel == "City Hotel" and is_canceled == 1') \
    .groupby('arrival_date_year') \
    .arrival_date_month \
    .value_counts()

arrival_date_year  arrival_date_month
2015               September             1543
                   October               1321
                   August                1232
                   July                   939
                   December               668
                   November               301
2016               October               1947
                   June                  1720
                   September             1567
                   April                 1539
                   May                   1436
                   November              1360
                   August                1247
                   March                 1108
                   December              1072
                   July                  1043
                   February               930
                   January                438
2017               May                   2217
                   April                 1926
                   June                  1

***
Look at the numerical characteristics of three variables: `adults`, `children`, and `babies`. Which of them has the `highest average`?

In [15]:
df[['adults', 'children', 'babies']] \
    .describe() \
    .loc['mean']

adults      1.856403
children    0.103890
babies      0.007949
Name: mean, dtype: float64

Result shows that `adult` guests are the most frequent visitors: near `1.8`

***
Create a `total_kids` column by combining children and babies. What type of hotels on average are more popular with clients with children?

In [18]:
df['total_kids'] = df['children'] + df['babies']
df.total_kids

0         0.0
1         0.0
2         0.0
3         0.0
4         0.0
         ... 
119385    0.0
119386    0.0
119387    0.0
119388    0.0
119389    0.0
Name: total_kids, Length: 119390, dtype: float64

In [17]:
df \
    .groupby('hotel') \
    .aggregate({'total_kids': 'mean'})

total_kids
hotel                   
City Hotel      0.096311
Resort Hotel    0.142586

As we can see, `Resort Hotel` is the most popular among guests with kids.

***
Create a `has_kids` variable that accepts the `True` value if the client specifies `at least one child` (total_kids) when booking, otherwise - `False`. 
Calculate the ratio of the number of retired users to the total number of clients, expressed as a percentage (`churn rate`). Indicate which group has the highest score.

In [21]:
df['has_kids'] = df['total_kids'] > 0

In [58]:
df[['has_kids']].sample(n=10)

has_kids
53626       True
118511     False
16542      False
59314      False
48351      False
65185      False
14036      False
27228       True
4425       False
71451      False

In [31]:
retire_clients_has_kids_num = df.query('has_kids == True and is_canceled == 1').shape[0]
total_clients_has_kids_num = df.query('has_kids == True').shape[0]

print(f'{retire_clients_has_kids_num=}')
print(f'{total_clients_has_kids_num=}')

churn_rate_has_kids = round(retire_clients_has_kids_num / total_clients_has_kids_num * 100, 2)
print(f'{churn_rate_has_kids=}')

retire_clients_has_kids_num=3259
total_clients_has_kids_num=9332
churn_rate_has_kids=34.92


In [32]:
retire_clients_no_kids_num = df.query('has_kids == False and is_canceled == 1').shape[0]
total_clients_no_kids_num = df.query('has_kids == False').shape[0]

print(f'{retire_clients_no_kids_num=}')
print(f'{total_clients_no_kids_num=}')

churn_rate_no_kids = round(retire_clients_no_kids_num / total_clients_no_kids_num * 100, 2)
print(f'{churn_rate_no_kids=}')

retire_clients_no_kids_num=40965
total_clients_no_kids_num=110058
churn_rate_no_kids=37.22


As calculations show, `churn rate` among clients that have no kids - `37.22` is more than one among clients with kids - `34.92`.